File: internal/agent/specgen.go
package agent

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	aikit "github.com/Volpestyle/ai-kit/packages/go"
	"github.com/yourorg/coding-agent-harness/internal/config"
)

type SpecGenerator struct {
	Kit    *aikit.Kit
	Router *aikit.ModelRouter
	Policy config.ModelPolicy
}

func (g *SpecGenerator) GenerateSpec(ctx context.Context, workspacePath, specName, prompt string) (string, error) {
	if g.Kit == nil {
		return "", errors.New("kit is nil")
	}
	model, err := g.resolveModel(ctx)
	if err != nil {
		return "", err
	}
	agents, _ := os.ReadFile(filepath.Join(workspacePath, "AGENTS.md"))
	sys := buildSpecPrompt(specName, prompt, string(agents))

	out, err := g.Kit.Generate(ctx, aikit.GenerateInput{
		Provider: model.Provider,
		Model:    model.ProviderModelID,
		Messages: []aikit.Message{{
			Role: "user",
			Content: []aikit.ContentPart{{
				Type: "text",
				Text: sys,
			}},
		}},
	})
	if err != nil {
		return "", err
	}
	content := strings.TrimSpace(out.Text)
	if content == "" {
		return "", errors.New("model returned empty spec")
	}
	if !strings.Contains(content, "# Goal") {
		content = fallbackSpec(specName, prompt)
	}
	if !strings.HasSuffix(content, "\n") {
		content += "\n"
	}
	return content, nil
}

func (g *SpecGenerator) resolveModel(ctx context.Context) (aikit.ModelRecord, error) {
	records, err := g.Kit.ListModelRecords(ctx, nil)
	if err != nil {
		return aikit.ModelRecord{}, err
	}
	if g.Router == nil {
		g.Router = &aikit.ModelRouter{}
	}
	resolved, err := g.Router.Resolve(records, aikit.ModelResolutionRequest{
		Constraints: aikit.ModelConstraints{
			RequireTools:  g.Policy.RequireTools,
			RequireVision: g.Policy.RequireVision,
			MaxCostUSD:    g.Policy.MaxCostUSD,
		},
		PreferredModels: g.Policy.PreferredModels,
	})
	if err != nil {
		return aikit.ModelRecord{}, err
	}
	return resolved.Primary, nil
}

func buildSpecPrompt(name, prompt, agents string) string {
	var b strings.Builder
	b.WriteString("You are an expert product/spec writer for a coding agent harness.\n")
	b.WriteString("Return ONLY markdown (no code fences, no commentary).\n")
	b.WriteString("Follow this exact structure:\n")
	b.WriteString("---\n")
	b.WriteString("name: " + name + "\n")
	b.WriteString("owner: you\n")
	b.WriteString("status: draft\n")
	b.WriteString("---\n\n")
	b.WriteString("# Goal\n\n")
	b.WriteString("<one paragraph goal>\n\n")
	b.WriteString("# Constraints / nuances\n\n")
	b.WriteString("- <bullets>\n\n")
	b.WriteString("# Acceptance tests\n\n")
	b.WriteString("- <bulleted, runnable checks>\n\n")
	b.WriteString("# Notes\n\n")
	b.WriteString("- <optional>\n\n")
	b.WriteString("USER PROMPT:\n")
	b.WriteString(prompt)
	b.WriteString("\n\n")
	if strings.TrimSpace(agents) != "" {
		b.WriteString("AGENTS.md:\n")
		b.WriteString(agents)
		b.WriteString("\n\n")
	}
	return b.String()
}

func fallbackSpec(name, prompt string) string {
	return fmt.Sprintf(`---
name: %s
owner: you
status: draft
---

# Goal

%s

# Constraints / nuances

- Follow repo conventions in AGENTS.md.

# Acceptance tests

- make test
`, name, strings.TrimSpace(prompt))
}

File: internal/agent/specgen.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import type { Kit, ModelRecord } from "@volpestyle/ai-kit-node";
import { ModelRouter } from "@volpestyle/ai-kit-node";
import type { ModelPolicy } from "../config";

export class SpecGenerator {
  constructor(
    private kit: Kit,
    private policy: ModelPolicy,
    private router: ModelRouter = new ModelRouter(),
  ) {}

  async generateSpec(workspacePath: string, specName: string, prompt: string): Promise<string> {
    if (!this.kit) throw new Error("kit is nil");
    const model = await this.resolveModel();
    let agents = "";
    try {
      agents = await readFile(path.join(workspacePath, "AGENTS.md"), "utf8");
    } catch {
      // ignore
    }
    const sys = buildSpecPrompt(specName, prompt, agents);

    const out = await this.kit.generate({
      provider: model.provider,
      model: model.providerModelId,
      messages: [
        {
          role: "user",
          content: [{ type: "text", text: sys }],
        },
      ],
    });
    let content = (out.text ?? "").trim();
    if (!content) throw new Error("model returned empty spec");
    if (!content.includes("# Goal")) {
      content = fallbackSpec(specName, prompt);
    }
    if (!content.endsWith("\n")) content += "\n";
    return content;
  }

  private async resolveModel(): Promise<ModelRecord> {
    const records = await this.kit.listModelRecords();
    const resolved = this.router.resolve(records, {
      constraints: {
        requireTools: this.policy.require_tools,
        requireVision: this.policy.require_vision,
        maxCostUsd: this.policy.max_cost_usd,
      },
      preferredModels: this.policy.preferred_models,
    });
    return resolved.primary;
  }
}

function buildSpecPrompt(name: string, prompt: string, agents: string): string {
  let out = "";
  out += "You are an expert product/spec writer for a coding agent harness.\n";
  out += "Return ONLY markdown (no code fences, no commentary).\n";
  out += "Follow this exact structure:\n";
  out += "---\n";
  out += `name: ${name}\n`;
  out += "owner: you\n";
  out += "status: draft\n";
  out += "---\n\n";
  out += "# Goal\n\n";
  out += "<one paragraph goal>\n\n";
  out += "# Constraints / nuances\n\n";
  out += "- <bullets>\n\n";
  out += "# Acceptance tests\n\n";
  out += "- <bulleted, runnable checks>\n\n";
  out += "# Notes\n\n";
  out += "- <optional>\n\n";
  out += "USER PROMPT:\n" + prompt + "\n\n";
  if (agents.trim()) {
    out += "AGENTS.md:\n" + agents + "\n\n";
  }
  return out;
}

function fallbackSpec(name: string, prompt: string): string {
  return `---\nname: ${name}\nowner: you\nstatus: draft\n---\n\n# Goal\n\n${prompt.trim()}\n\n# Constraints / nuances\n\n- Follow repo conventions in AGENTS.md.\n\n# Acceptance tests\n\n- make test\n`;
}
